### ISSUE TITLE
Callable returned from factory and assigned to class attribute is not treated as a bound method

### ISSUE THREAD

**Issue by Hannes237** [astral-sh/ty#4074]
## Summary

Ty appears to incorrectly handle functions returned from a factory when assigned to a class attribute.

The callable should behave as a normal bound method when accessed through an instance, but Ty instead reports that the first parameter is missing.

## Minimal reproduction

```python
from collections.abc import Callable

def unary() -> Callable[["Foo"], int]:
    def f(self: "Foo") -> int:
        return 42
    return f

class Foo:
    bar = unary()

Foo().bar()
reveal_type(unary())
reveal_type(Foo().bar)
```
Resulting in 
```
    |
    | reveal_type(Foo().bar)
    |             ^^^^^^^^^ `(Foo, /) -> int`
```

### Version

ty 0.0.62 (df75992e7 2026-07-21)

**Comment by carljm** [comment]
This is #491. The Python type system does not provide enough information to decide whether a callable type should be treated as a bound-method descriptor or not. The return annotation `Callable[["Foo"], int]` does not clarify: you could return either a function or a callable instance which is not a bound-method descriptor, and both would be assignable to that return type. So type checkers have to use heuristics to guess which is more likely, and are guaranteed to get some cases wrong (and all type checkers do). That said, I think we could improve our heuristics here, and our default should probably be to just match the heuristics used by pyright/mypy. Which I've worked on a bit in astral-sh/ruff#26778.

Thanks for the report!
